File: src/docs/website/site/chromium-os/developer-library/guides/testing/web-testing/pwa/test-pwa-lifecycle.py
# Copyright 2026 The Chromium Authors
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Selenium Test Script for a Standard PWA Lifecycle.

This script connects to a remote Chrome instance, installs a standard
PWA using the `PWA` Chrome DevTools Protocol (CDP) domain. It then
launches the PWA, checks its state, and finally uninstalls it.

Prerequisites:
1. A CrOS device or VM setup for remote testing as described in the
   guide:
   https://www.chromium.org/chromium-os/developer-library/guides/testing/web-testing/
2. Chrome on the device must be launched with the following flags in
   /etc/chrome_dev.conf:
   - `--remote-debugging-port=<port>`
   - `--enable-devtools-pwa-handler`
"""

import argparse
import logging
import sys
import time

from selenium import webdriver
from selenium.webdriver.chrome import options
from selenium.webdriver.chrome import service
from selenium.webdriver.support import ui

# Configure logging.
logging.basicConfig(level=logging.INFO, format="[%(levelname)s] %(message)s")

# PWA specific constants.
_MANIFEST_ID = "https://developer.chrome.com/"


def run_pwa_lifecycle_test(driver_path: str, port: int) -> bool:
    """Connects to a remote browser, installs, launches, and inspects a PWA.

    Args:
        driver_path: Path to the chromedriver binary.
        port: The remote-debugging-port flag value on the DUT.

    Returns:
        True on success, False on error.
    """
    debugger_addr = f"127.0.0.1:{port}"
    logging.info("Attempting to connect to browser at %s...", debugger_addr)

    chrome_opts = options.Options()
    chrome_opts.add_experimental_option("debuggerAddress", debugger_addr)

    chrome_svc = service.Service(executable_path=driver_path)

    try:
        with webdriver.Chrome(service=chrome_svc,
                              options=chrome_opts) as driver:
            logging.info("Successfully connected to the browser.")
            original_window = driver.current_window_handle
            wait = ui.WebDriverWait(driver, 15)

            # Install PWA.
            logging.info("Installing PWA from manifest: %s", _MANIFEST_ID)
            driver.execute_cdp_cmd(
                "PWA.install",
                {
                    "manifestId": _MANIFEST_ID,
                    "installUrlOrBundleUrl": _MANIFEST_ID,
                },
            )
            # Allow time for installation to propagate.
            time.sleep(5)

            # Launch PWA.
            logging.info("Launching PWA with manifest ID: %s", _MANIFEST_ID)
            driver.execute_cdp_cmd("PWA.launch", {"manifestId": _MANIFEST_ID})

            # Wait for the PWA window to open.
            logging.info("Waiting for PWA window to open...")
            wait.until(lambda d: len(d.window_handles) > 1)

            # Find new handle.
            pwa_window_handle = next(
                (h for h in driver.window_handles if h != original_window),
                None)

            if not pwa_window_handle:
                raise RuntimeError("PWA window handle not found.")

            driver.switch_to.window(pwa_window_handle)
            # Ensure we are on the right URL (sometimes launch opens blank).
            driver.get(_MANIFEST_ID)
            logging.info("PWA launched successfully and window is focused.")

            # Inspect state.
            logging.info("Inspecting PWA's OS state and manifest...")
            os_app_state = driver.execute_cdp_cmd("PWA.getOsAppState",
                                                  {"manifestId": _MANIFEST_ID})
            logging.info("PWA OS App State: %s", os_app_state)

            manifest = driver.execute_cdp_cmd("Page.getAppManifest", {})
            logging.info("PWA Manifest: %s", manifest)

            # Test opening current page.
            logging.info("Testing 'openCurrentPageInApp' command...")
            driver.execute_cdp_cmd("PWA.openCurrentPageInApp",
                                   {"manifestId": _MANIFEST_ID})

            # Cleanup.
            try:
                logging.info("Uninstalling PWA: %s", _MANIFEST_ID)
                driver.execute_cdp_cmd("PWA.uninstall",
                                       {"manifestId": _MANIFEST_ID})
            except Exception as e:
                logging.warning("Failed to uninstall PWA during cleanup: %s", e)

            logging.info("Test complete.")
            return True

    except Exception as e:
        logging.error("An error occurred during the test: %s", e)
        return False


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--driver-path",
        default="./chromedriver",
        help="Path to the chromedriver binary.",
    )
    parser.add_argument("--port",
                        type=int,
                        default=9222,
                        help="Remote debugging port.")
    opts = parser.parse_args(argv)

    return 0 if run_pwa_lifecycle_test(opts.driver_path, opts.port) else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/docs/website/site/chromium-os/developer-library/guides/testing/web-testing/pwa/test-pwa-lifecycle.ts
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Playwright Test Script for a Standard PWA Lifecycle
 * (TypeScript).
 *
 * This script connects to a remote Chrome instance, installs a standard PWA
 * using the `PWA` Chrome DevTools Protocol (CDP) domain. It then launches the
 * PWA, checks its state, and finally uninstalls it.
 *
 * Prerequisites:
 * 1. A CrOS device or VM setup for remote testing.
 * 2. Chrome must be launched with:
 * - `--remote-debugging-port=<port>`
 * - `--enable-devtools-pwa-handler`
 */

import * as fs from 'fs';
import * as path from 'path';
import {Browser, BrowserContext, CDPSession, chromium, Page} from 'playwright';

// PWA specific constants.
const PWA_MANIFEST_ID = 'https://developer.chrome.com/';

// Color constants for console output.
const COLORS = {
  normal: '\x1b[m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
};

// Helper for colorless environments.
if (process.stderr && !process.stderr.hasColors()) {
  (Object.keys(COLORS) as (keyof typeof COLORS)[]).forEach(
      (key) => COLORS[key] = '');
}

/**
 * Helper to send CDP commands with logging.
 */
async function sendCDPCommand(
    session: CDPSession, command: string, params: object): Promise<any> {
  console.log(COLORS.green, `Sending CDP command: ${command}`, COLORS.normal);
  try {
    const result = await session.send(command as any, params as any);
    // Log response, trimmed if too long.
    const resultStr = JSON.stringify(result);
    console.log(`Response: ${
        resultStr.length > 200 ? resultStr.substring(0, 200) + '...' :
                                 resultStr}`);
    return result;
  } catch (e: any) {
    console.error(
        COLORS.red, `CDP Error (${command}): ${e.message}`, COLORS.normal);
    throw e;
  }
}

async function runPWALifecycleTest() {
  const port = process.argv[2] || '9222';
  const browserURL = `http://localhost:${port}`;

  let browser: Browser|null = null;
  let context: BrowserContext|null = null;
  let managementPage: Page|null = null;
  let pwaPage: Page|null = null;

  try {
    console.log(`Attempting to connect to browser at ${browserURL}...`);
    browser = await chromium.connectOverCDP(browserURL);
    console.log('Successfully connected.');

    // Get default context or create one if missing.
    if (browser.contexts().length > 0) {
      context = browser.contexts()[0];
    } else {
      context = await browser.newContext();
    }

    // 1. Create a session to perform PWA management commands.
    // We open a blank page to attach the CDP session to.
    managementPage = await context.newPage();
    const client = await context.newCDPSession(managementPage);
    console.log('Created CDP session for PWA management.');

    const manifestParam = {manifestId: PWA_MANIFEST_ID};
    const installParam = {
      ...manifestParam,
      installUrlOrBundleUrl: PWA_MANIFEST_ID,
    };

    // 2. Install the PWA.
    console.log('Installing PWA from:', installParam.installUrlOrBundleUrl);
    await sendCDPCommand(client, 'PWA.install', installParam);
    console.log('PWA installation command sent.');

    // 3. Launch the PWA.
    console.log('Launching PWA with manifest ID:', manifestParam.manifestId);
    await sendCDPCommand(client, 'PWA.launch', manifestParam);

    // 4. Wait for the PWA window to open.
    console.log('Waiting for PWA window to open... (Timeout: 15 seconds)');

    // Playwright waits for the 'page' event on the context.
    pwaPage = await context.waitForEvent('page', {
      predicate: (page) => page.url().startsWith(PWA_MANIFEST_ID),
      timeout: 15000
    });

    await pwaPage.bringToFront();
    console.log('PWA launched successfully and window is focused.');

    // 5. Inspect the PWA's state and manifest.
    console.log('Inspecting PWA\'s OS state and manifest...');
    await sendCDPCommand(client, 'PWA.getOsAppState', manifestParam);

    // Create a specific CDP session for the PWA page itself to query its
    // specific page manifest.
    const pwaSession = await context.newCDPSession(pwaPage);
    await sendCDPCommand(pwaSession, 'Page.getAppManifest', {});

    // 6. Test opening the current page in its app window.
    console.log('Testing \'openCurrentPageInApp\' command...');
    await sendCDPCommand(pwaSession, 'PWA.openCurrentPageInApp', manifestParam);

    console.log('Test complete. Proceeding to uninstall.');

  } catch (error: any) {
    console.error(
        COLORS.red, 'An error occurred during the test:', error.message,
        COLORS.normal);

    // Screenshot logic.
    if (pwaPage) {
      try {
        const screenshotDir = path.join(__dirname, 'screenshots');
        if (!fs.existsSync(screenshotDir)) {
          fs.mkdirSync(screenshotDir);
        }
        const screenshotPath =
            path.join(screenshotDir, 'failed_pwa_test_ts.png');
        await pwaPage.screenshot({path: screenshotPath});
        console.log(
            COLORS.yellow, `Screenshot saved to ${screenshotPath}`,
            COLORS.normal);
      } catch (screenshotError) {
        console.error('Failed to take screenshot:', screenshotError);
      }
    }
    process.exitCode = 1;
  } finally {
    console.log('Cleanup Phase...');
    // 7. Uninstall PWA and shutdown.
    if (browser && context) {
      // We use a valid page session for uninstallation.
      // If managementPage is closed, create a temporary one.
      let cleanupPage = managementPage;
      if (!cleanupPage || cleanupPage.isClosed()) {
        try {
          cleanupPage = await context.newPage();
        } catch (e) { /* ignore */
        }
      }

      if (cleanupPage && !cleanupPage.isClosed()) {
        console.log('Attempting to uninstall PWA...');
        try {
          const client = await context.newCDPSession(cleanupPage);

          // Close the PWA window first if it's open.
          if (pwaPage && !pwaPage.isClosed()) {
            console.log('Closing PWA window before uninstalling...');
            await pwaPage.close();
          }

          await sendCDPCommand(
              client, 'PWA.uninstall', {manifestId: PWA_MANIFEST_ID});
          console.log('PWA uninstalled.');
        } catch (e: any) {
          console.warn(
              COLORS.yellow,
              'Warning: Failed to uninstall PWA during cleanup:', e.message,
              COLORS.normal);
        }

        // Only close if we created a temp page or it's the original management
        // page.
        if (cleanupPage) await cleanupPage.close();
      }

      console.log('Disconnecting from browser...');
      await browser.close();
      console.log('Shutdown complete.');
    }
  }
}

runPWALifecycleTest();
